trace_decoder/src/decoding.rs

```diff
@@ -302,11 +295,122 @@ impl ProcessedBlockTrace {
         Ok(txn_gen_inputs)
     }
 
+    /// Cancun HF specific: At the start of a block, prior txn execution, we
+    /// need to update the storage of the beacon block root contract.
+    // See <https://eips.ethereum.org/EIPS/eip-4788>.
+    fn update_beacon_block_root_contract_storage(
+        trie_state: &mut PartialTrieState,
+        delta_out: &mut TrieDeltaApplicationOutput,
+        nodes_used: &mut NodesUsedByTxn,
+        block_data: &BlockMetadata,
+    ) -> TraceParsingResult<()> {
+        const HISTORY_BUFFER_LENGTH_MOD: U256 = U256([HISTORY_BUFFER_LENGTH.1, 0, 0, 0]);
+        const ADDRESS: H256 = H256(BEACON_ROOTS_CONTRACT_ADDRESS_HASHED);
+
+        let timestamp_idx = block_data.block_timestamp % HISTORY_BUFFER_LENGTH_MOD;
+        let timestamp = rlp::encode(&block_data.block_timestamp).to_vec();
+
+        let root_idx = timestamp_idx + HISTORY_BUFFER_LENGTH_MOD;
+        let calldata = rlp::encode(&block_data.parent_beacon_block_root).to_vec();
+
+        let storage_trie = trie_state
+            .storage
+            .get_mut(&ADDRESS)
+            .ok_or(TraceParsingError::new(
+                TraceParsingErrorReason::MissingAccountStorageTrie(ADDRESS),
+            ))?;
+
+        let mut slots_nibbles = vec![];
+
+        for (slot, val) in [(timestamp_idx, timestamp), (root_idx, calldata)]
+            .iter()
+            .map(|(k, v)| {
+                (
+                    Nibbles::from_h256_be(hash(
+                        &Nibbles::from_h256_be(H256::from_uint(k)).bytes_be(),
+                    )),
+                    v,
+                )
+            })
+        {
+            slots_nibbles.push(slot);
+
+            // If we are writing a zero, then we actually need to perform a delete.
+            match val == &ZERO_STORAGE_SLOT_VAL_RLPED {
+                false => {
+                    storage_trie.insert(slot, val.clone()).map_err(|err| {
+                        let mut e =
+                            TraceParsingError::new(TraceParsingErrorReason::TrieOpError(err));
+                        e.slot(U512::from_big_endian(slot.bytes_be().as_slice()));
+                        e.slot_value(U512::from_big_endian(val.as_slice()));
+                        e
+                    })?;
+
+                    delta_out
+                        .additional_storage_trie_paths_to_not_hash
+                        .entry(ADDRESS)
+                        .or_default()
+                        .push(slot);
```

Comment:
Yeah I originally wanted to, but it felt a bit heavy because of the return type of `delete_node_and_report_remaining_key_if_branch_collapsed()`.
I guess I could still make it work, but not sure it's worth the hassle.